### DISCUSSION TITLE
Windows binary

### DISCUSSION THREAD

**Discussion by vladdeSV** [jotaen/klog#386] [Feature ideas & discussions]
Compiling klog on Windows has no apparent problems when using Windows Terminal.

![klog windows terminal](https://user-images.githubusercontent.com/7542961/107976489-ee258000-6fb9-11eb-9296-6f7809cf7713.png)

Are there any reasons to why a Windows binary is not present?

**Comment by vladdeSV** [comment]
The only problem I could find from my limited testing is that it does not color text when using *conhost.exe*.

![klog windows powershell](https://user-images.githubusercontent.com/7542961/107976737-51afad80-6fba-11eb-8485-3120a7c6fde0.png)

**Comment by jotaen** [comment]
Oh yes, there is a pretty good reason: I have basically no idea about Windows. What I did was to give it a brief shot on a Windows 10 VM, but I saw something similar to your second screenshot and then I gave up more or less immediately 🙈 
I was also not sure whether the Windows kids nowadays actually use CLI-based productivity tools, but times might have changed.

Anyways, the first screenshot gives me some hope, can you tell me what the difference is between the first and the second? Maybe it’s just a matter of documentation.

As far as I see there should be only three potential issues with portability:

- The line endings
- The escape sequences for the colours (`\033[X;X;Xm`), though that’s actually more related to the shell rather than to the OS
- The bookmark functionality, which creates a dot-folder in your homefolder and puts a symlink in there. That’s all accessed through the go library and *should* be portable, but I haven’t confirmed that.

**Comment by vladdeSV** [comment]
The first screenshot is taken with [Windows Terminal](https://github.com/microsoft/terminal), which *works* with ANSI escape codes. For me, that whole program lies in the "magic" category. "Magic", as in I do not fully understand how it works.

The three points you mentioned are valid concerns, all of which I had not properly though about.

- Line endings in Windows default to CRLF. I had to manually change the line endings in VS Code to get it to parse properly.
- Escape sequences only work with Windows Terminal, WSL, or [Git Bash](https://github.com/git-for-windows/git) (bash as shell in Windows)
- If I remember correctly, Windows does not allow symlinks by default. I believe you have to enable symlinks in some setting somewhere. I am uncertain about this.

**Comment by vladdeSV** [comment]
The second screenshot was taken with PowerShell, using the default Windows shell `conhost.exe`.

Again, I do not fully understand how it all ties together.

**Comment by jotaen** [comment]
Awesome, thanks for the info!

- The line endings can just be detected and converted by the klog parser via regex
- That’s good info that there exists an ANSI compatible terminal. I wasn’t even aware that there is something other than Powershell and cmd.exe 😆 
- Go’s `os.Symlink` should work on Windows (https://codereview.appspot.com/86160044), but it’s a good hint to check with the system settings.

With these things done I think there is no reason anymore to not offer the Win binary for downloading 😸

**Comment by jotaen** [comment]
Windows line endings support as of https://github.com/jotaen/klog/pull/21 (not released yet)
